Feature/get user info

randychoi95 opened the pull request:
# Get UserInfo
- 키 체인에 저장된 데이터(ASAuthorizationAppleIDCredential의 user, UserInfo의 id)를 이용하여 Realm 저장소에 저장된 사용자 정보 가져오기
- AppRoot Riblet에서 이전에 로그인한 적이 체크
- 로그인한 적이 있는 경우: 사용자 정보를 불러오기
- 로그인한 적이 없는 경우: LoggedOut Riblet attach 하기

# 다음 진행사항
- 로그인한 후 Riblet 생성 및 구현

randychoi95 committed:
:sparkles: 키체인 저장소에서 저장된 데이터 가져오기

- 키체인 저장할 때 인자값 순서 바꿈(-> key, value가 바뀌어서 설정이 되어있었음)
- 키체인에서 데이터 가져오기 성공 -> 사용자 정보 가져오기 실행
- 키체인에서 데이터 가져오기 실패 -> 로그인 화면으로 이동

randychoi95 committed:
:sparkles: Realm에 저장된 사용자 정보 가져오기

- RealmDatabase의 read메소드의 인자 값(key) 추가(-> key를 통해 저장된 데이터 가져오기)
- 키체인 저장소에 realm객체의 id(ObjectId) 저장
- 키체인에 저장된 id(ObjectId)로 Realm에 저장된 데이터(사용자 정보) 가져오기

f-lab-max commented on PlaceStory/Domain/Sources/UseCase/AppleAuthenticationServiceUseCase.swift:
```diff
@@ -8,10 +8,13 @@
 import Combine
 import Entities
 import Foundation
+import RealmSwift
```
필요한 import인가요? 쓰는 곳이 없어보여서요

randychoi95 replied:
이 부분 삭제하겠습니다

f-lab-max commented on PlaceStory/Platform/Sources/LocalStorage/Realm/RealmDatabase.swift:
```diff
@@ -24,7 +24,7 @@ enum RealmDatabaseError: Error {
 
 public protocol RealmDatabase {
```
데이터베이스가 Realm 이라는 것은 구현 디테일이지, 인터페이스에 그것이 드러날 필요는 없는거 같습니다. 혹시 Realm을 드러내는 다른 이유가 있나요?

randychoi95 replied:
아니요. 드러내지 않게 하겠습니다.

f-lab-max commented on PlaceStory/Platform/Sources/RepositoryImps/AppleAuthenticationServiceRepositoryImp.swift:
```diff
@@ -74,44 +164,24 @@ extension AppleAuthenticationServiceRepositoryImp: ASAuthorizationControllerDele
+            signInSubject.send(userInfo.toDomain())
         } else {
-            Log.error(createResult.resultMessage, "[\(#file)-\(#function) - \(#line)]")
+            RealmDatabaseImp.shared.create(userInfo)
```
데이터베이스는 대표적인 volatile dependency로, 의존성 주입의 대상입니다. 직접 참조하여 사용하지 말고 주입받아서 느슨한 결합을 달성하여 테스트 가능한 코드가 되면 좋겠습니다.

randychoi95 replied:
직접 참조하지 않고 느슨한 결합을 통하여 테스트 가능한 코드가 되게 수정해보겠습니다

randychoi95 replied:
해당 내용에 대해서 반영해보았습니다.

randychoi95 committed:
:zap: 리뷰 반영

- 불필요한 import 및 접근제한자 삭제
- RealmDatabase에 대한 느슨한 결합을 달성하게 수정
- 추가적으로 KeyChainService에 대해서도 느슨한 결합을 달성하게 수정

f-lab-max commented on PlaceStory/AppRoot/Sources/AppRoot/AppRootInteractor.swift:
```diff
@@ -28,12 +31,17 @@ final class AppRootInteractor: PresentableInteractor<AppRootPresentable>, AppRoo
     init(
         presenter: AppRootPresentable,
-        appleAuthenticationServiceUseCase: AppleAuthenticationServiceUseCase
+        appleAuthenticationServiceUseCase: AppleAuthenticationServiceUseCase,
+        appleAuthenticationServiceRepository: AppleAuthenticationServiceRepository
```
usecase와 repository의 차이는 무엇인가요? 앞으로 생길 모든 usecase에 대해 repository도 별도로 존재하게 되는건가요? interactor가 둘다 의존해야하나요?

randychoi95 replied:
Usecase는 앱의 비즈니스 로직을 정의하고 구현하는 곳입니다. 이에 대비해 repository는 데이터를 가져오고 저장하는 역할을 수행합니다. 각 usecase마다 별도의 repository를 사용할수도 있고, 동일한 repository를 공유하여 사용할 수도 있습니다.

현재 상황에서는 애플로그인 이전에 사용한 적이 있는지 판단하기 위해 usecase를 활용하고 있습니다. repository에서 사용자 정보에 대한 변수를 선언하고 사용하려고 하였지만, usecase에서 직접 사용자 정보를 가져올 수 있도록 구현해서 repository에 대한 의존하지 않아도 됩니다.
